Descriptor expansion cache clarifications

Sjors opened the pull request:
I found the name `m_script_arg` to be confusing while reviewing https://github.com/bitcoin/bitcoin/pull/14646#discussion_r240677238. @sipa let me know if `m_subdescriptor_arg` is completely wrong.

I also added an explanation of why we call `GetPubKey` when we don't ask it for a public key.

Sjors committed:
scripted-diff: rename DescriptorImpl m_script_arg to m_subdescriptor_arg

-BEGIN VERIFY SCRIPT-
sed -i -e 's/m_script_arg/m_subdescriptor_arg/g' src/script/descriptor.cpp
-END VERIFY SCRIPT-

Sjors committed:
[doc] descriptor: explain GetPubKey() usage with cached public key

Plus a few typo fixes.

fanquake commented:
Might as well fixup the other nit (trivial) [nits](https://github.com/bitcoin/bitcoin/pull/14646#pullrequestreview-183028560).

Sjors commented:
@fanquake done, I snuck those into the second commit.

flack commented on src/script/descriptor.cpp:
```diff
@@ -295,6 +295,8 @@ class DescriptorImpl : public Descriptor
         for (const auto& p : m_pubkey_args) {
             entries.emplace_back();
+            // If we have a cache, GetPubKey should not actually return a public key,
+            // so we pass a nullprt. We're only interested in uncached origin info.
```
Typo: `nullprt`

Sjors replied:
Fixed

MeshCollider commented:
`m_script_arg` refers to the defined `SCRIPT` in the descriptor doc, to make it distinct from `KEY` or `ADDR`. But subdescriptor is also a clear name so I have no strong preference either way.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -295,6 +295,8 @@ class DescriptorImpl : public Descriptor
         // Construct temporary data in `entries` and `subscripts`, to avoid producing output in case of failure.
         for (const auto& p : m_pubkey_args) {
             entries.emplace_back();
+            // If we have a cache, GetPubKey should not actually return a public key,
```
"should not" sounds like it should be prohibited from computing the pubkey.

What about "If we have a cache, we don't need GetPubKey to compute the public key. Pass in nullptr to signify only origin info is desired.".

Sjors replied:
Done

Sjors commented:
Rebased.

@MeshCollider I added a comment referencing `SCRIPT` in the descriptor docs.

By the way, I noticed that neither `DescriptorImpl` nor `m_subdescriptor_arg` show up in the docs, not even in the docs for `script/descriptor.cpp`. I don't understand Doxygen well enough to know if that's a feature (i.e. because these functions are private).

Sjors commented:
Miraculously this doesn't need another rebase...

practicalswift commented:
FWIW: I've verified that a disassembly of the `bitcoind` binary built with this patch applied is identical to a disassembly of the `bitcoind` binary built against `master` (as expected).

laanwj commented:
ACK 2e68ffaf205866e4cea71f64e79bbfb89e17280a
